### PR TITLE
Adding finalize invite features

### PR DESCRIPTION
**Pull request by kayladavis**
Admins now get a "finalize invites" button that pops up a modal for confirmation.
Students see pending until invites have been finalized.

image

image

### CODE REVIEW

**Comment by kayladavis** [comment]
I didn't realize that #213 was a separate story. This PR includes #213 .

File: backend/db/migrations/20180329230713-create-show.js [hunk 1]
```diff
@@ -37,6 +37,11 @@ export function up (queryInterface, Sequelize) {
       type: Sequelize.DATE,
       allowNull: false
     },
+    finalized: {
```
Comment (billdybas) [inline backend/db/migrations/20180329230713-create-show.js]: Don't update the timestamp on this file. You break migrations since there's now a `user_shows` migration before a `shows` migration.

Comment (billdybas) [reply backend/db/migrations/20180329230713-create-show.js]: Ideally this would have been a new migration w/ [`addColumn`](http://docs.sequelizejs.com/class/lib/query-interface.js~QueryInterface.html#instance-method-addColumn), but whatever...

File: backend/schema.js [hunk 2]
```diff
@@ -62,13 +63,14 @@ type Show {
 }
 
 input ShowInput {
-    name: String!
+    name: String
```
Comment (billdybas) [inline backend/schema.js]: Why did you remove the required checks?

File: frontend/src/Student/components/ShowCard.js [hunk 3]
```diff
@@ -134,17 +134,18 @@ const SubmittedEntries = ({ show }) =>
...
-            <JudgingPhase>Judging In Progress</JudgingPhase>
-          ) : moment().isAfter(show.judgingEnd) ? (
-            entry.invited ? (
-              <Accepted>Invited</Accepted>
-            ) : (
-              <NotAccepted>Not Invited</NotAccepted>
-            )
-          ) : null}
+        {/* If after entry end and before judging end (or if the show is not finalized),
+          display "judging phase", else display accepted or denied */
```
Comment (billdybas) [inline frontend/src/Student/components/ShowCard.js]: Update this comment.

File: frontend/src/Admin/mutations/finalizeShowInvites.graphql [hunk 4]
```diff
@@ -0,0 +1,6 @@
+mutation finalizeShowInvites($id: ID! $input: ShowInput!) {
+  updateShow(id: $id, input: $input) {
+    id,
+    finalized
+  }
+}
```
Comment (billdybas) [inline frontend/src/Admin/mutations/finalizeShowInvites.graphql]: nit: new line

If you're using VSCode, might I suggest adding these to your editor's config:

```
"files.trimTrailingWhitespace": true,
"files.insertFinalNewline": true,
"files.trimFinalNewlines": true,
```

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 5]
```diff
@@ -45,7 +52,18 @@ class ShowSubmissionsTab extends Component {
 
     this.state = {
       alertVisible: true,
-      alertType: ''
+      alertType: '',
+      showFinalized: false,
+      modal: false
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: `isModalOpen`

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 6]
```diff
@@ -79,6 +97,24 @@ class ShowSubmissionsTab extends Component {
       })
   }
 
+  finalizeShowInvites = id => {
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: I don't think you need to make a new function for this - just add it above.

File: frontend/src/Admin/containers/ShowSubmissionsTab.js [hunk 7]
```diff
@@ -17,5 +18,18 @@ export default compose(
           }
         })
     })
+  }),
+  graphql(FinalizeShowInvites, {
+    props: ({ mutate }) => ({
+      finalizeInvites: id =>
```
Comment (billdybas) [inline frontend/src/Admin/containers/ShowSubmissionsTab.js]: Instead of passing around the `id` all over the place, you could probably get it from react-router in `ownProps`

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 8]
```diff
@@ -114,6 +150,35 @@ class ShowSubmissionsTab extends Component {
...
+          <ModalBody>
+            This is a permanent action. Clicking continue will make the invite
+            status for all submissions viewable to students.
+          </ModalBody>
+          <ModalFooter>
+            <Button
+              color='primary'
+              onClick={() => this.toggleFinalizeInviteModal(show.id, true)}
+            >
+              Make invites public
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: "Continue"

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 9]
```diff
@@ -114,6 +150,35 @@ class ShowSubmissionsTab extends Component {
...
           There was an error updating the invite status
         </Alert>
+        <Modal isOpen={this.state.modal}>
+          <ModalBody>
+            This is a permanent action. Clicking continue will make the invite
+            status for all submissions viewable to students.
+          </ModalBody>
+          <ModalFooter>
+            <Button
+              color='primary'
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: `color='danger'`

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 10]
```diff
@@ -15,17 +15,24 @@ import { getImageThumbnail } from '../../utils'
 const SUCCESS = 'SUCCESS'
 const ERROR = 'ERROR'
 
+// Invite button Message
+const FINALIZE_INVITES = 'Make invites public'
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: I don't think you really need to make constants for these.

And I think they should word:
"Make Invitations Public"
"Invitations Are Public"

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 11]
```diff
@@ -114,6 +150,35 @@ class ShowSubmissionsTab extends Component {
...
+            status for all submissions viewable to students.
+          </ModalBody>
+          <ModalFooter>
+            <Button
+              color='primary'
+              onClick={() => this.toggleFinalizeInviteModal(show.id, true)}
+            >
+              Make invites public
+            </Button>{' '}
+            <Button
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: Put this button on the left of the "Make Public" button.

File: backend/resolvers/mutations/show.js [hunk 12]
```diff
@@ -9,6 +9,17 @@ export function createShow (_, args, req) {
...
 }
 
+export function updateShow (_, args, req) {
+  // Only admins can update entries
+  if (req.auth.type !== ADMIN) {
+    throw new UserError('Permission Denied')
+  }
+  return Show.findById(args.id)
+    .then((show) => {
+      return show.update(args.input)
```
Comment (billdybas) [inline backend/resolvers/mutations/show.js]: You should only add update the columns you expect – [`options.fields`](http://docs.sequelizejs.com/class/lib/model.js~Model.html#instance-method-save) is what you want

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 13]
```diff
@@ -45,7 +52,18 @@ class ShowSubmissionsTab extends Component {
 
     this.state = {
       alertVisible: true,
-      alertType: ''
+      alertType: '',
+      showFinalized: false,
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: I don't think you need this?

Shouldn't you be able to get the updated value from `props`?

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 14]
```diff
@@ -79,6 +97,24 @@ class ShowSubmissionsTab extends Component {
...
+        showFinalized: true
+      })
+    })
+  }
+
+  componentDidMount () {
+    const { show } = this.props
+    if (show.finalized) {
+      this.setState({
+        showFinalized: true
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: See above - this is redundant

File: frontend/src/Admin/components/ShowSubmissionsTab.js [hunk 15]
```diff
@@ -114,6 +150,35 @@ class ShowSubmissionsTab extends Component {
         >
           There was an error updating the invite status
         </Alert>
+        <Modal isOpen={this.state.modal}>
+          <ModalBody>
+            This is a permanent action. Clicking continue will make the invite
```
Comment (billdybas) [inline frontend/src/Admin/components/ShowSubmissionsTab.js]: "This is a permanent action and will make invitations for this show visible to all students."

File: backend/resolvers/mutations/show.js [hunk 16]
```diff
@@ -9,6 +9,17 @@ export function createShow (_, args, req) {
   return Show.create(args.input)
 }
 
+export function updateShow (_, args, req) {
```
Comment (billdybas) [inline backend/resolvers/mutations/show.js]: Do we want to allow updating all the fields of a show right now? I feel like this could add in some edge cases w/ dates. Perhaps make a mutation just for finalizing since that's only what we need right now.

File: frontend/src/Student/components/ShowCard.js [hunk 17]
```diff
@@ -134,17 +134,18 @@ const SubmittedEntries = ({ show }) =>
...
+          moment().isBetween(show.entryEnd, show.judgingEnd) ||
+        !show.finalized ? (
+              <JudgingPhase>Pending</JudgingPhase>
```
Comment (billdybas) [inline frontend/src/Student/components/ShowCard.js]: Update this component name

File: frontend/src/Student/components/ShowCard.js [hunk 18]
```diff
@@ -134,17 +134,18 @@ const SubmittedEntries = ({ show }) =>
...
+            ) : moment().isAfter(show.judgingEnd) ? (
+              entry.invited ? (
+                <Accepted>Invited</Accepted>
```
Comment (billdybas) [inline frontend/src/Student/components/ShowCard.js]: Update this component name

File: frontend/src/Student/components/ShowCard.js [hunk 19]
```diff
@@ -134,17 +134,18 @@ const SubmittedEntries = ({ show }) =>
...
+              ) : (
+                <NotAccepted>Not Invited</NotAccepted>
```
Comment (billdybas) [inline frontend/src/Student/components/ShowCard.js]: Update this component name

File: backend/schema.js (hunk 2 above)
Comment (kayladavis) [reply backend/schema.js]: To allow for updates to work properly. It was too test stuff out, I'll create a ShowUpdate type.

File: backend/db/migrations/20180329230713-create-show.js (hunk 1 above)
Comment (kayladavis) [reply backend/db/migrations/20180329230713-create-show.js]: yeah, I know, but we've yet to do that. The process we've had since we started was to just update. I can go back to the old name and do the new add column thing.

File: backend/resolvers/mutations/show.js (hunk 16 above)
Comment (kayladavis) [reply backend/resolvers/mutations/show.js]: I think we should keep this mutation and instead just not allow for updating dates in a new ShowUpdate type (for now).
